clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,952 @@
+                // should just exit now.
+                if (state == ClientTelemetryState.TERMINATED || state == ClientTelemetryState.TERMINATING_PUSH_IN_PROGRESS) {
+                    return Optional.empty();
+                }
+
+                // Check the *actual* state (while locked) to make sure we're still in the state
+                // we expect to be in.
+                terminating = state == ClientTelemetryState.TERMINATING_PUSH_NEEDED;
+                if (!maybeSetState(terminating ? ClientTelemetryState.TERMINATING_PUSH_IN_PROGRESS : ClientTelemetryState.PUSH_IN_PROGRESS)) {
+                    return Optional.empty();
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+
+            byte[] payload;
+            try (MetricsEmitter emitter = new ClientTelemetryEmitter(localSubscription.selector(), localSubscription.deltaTemporality())) {
+                emitter.init();
+                kafkaMetricsCollector.collect(emitter);
+                payload = createPayload(emitter.emittedMetrics());
+            }
+
+            CompressionType compressionType = ClientTelemetryUtils.preferredCompressionType(localSubscription.acceptedCompressionTypes());
+            ByteBuffer buffer = ClientTelemetryUtils.compress(payload, compressionType);
+
+            AbstractRequest.Builder<?> requestBuilder = new PushTelemetryRequest.Builder(
+                new PushTelemetryRequestData()
+                    .setClientInstanceId(localSubscription.clientInstanceId())
+                    .setSubscriptionId(localSubscription.subscriptionId())
+                    .setTerminating(terminating)
+                    .setCompressionType(compressionType.id)
+                    .setMetrics(Utils.readBytes(buffer)), true);
+
+            return Optional.of(requestBuilder);
+        }
+
+        /**
+         * Updates the {@link ClientTelemetrySubscription}, {@link #intervalMs}, and
+         * {@link #lastRequestMs}.
+         * <p>
+         * The entire contents of the method are guarded by the {@link #lock}.
+         * <p>
+         * After the update, the {@link #subscriptionLoaded} condition is signaled so any threads
+         * waiting on the subscription can be unblocked.
+         *
+         * @param subscription Updated subscription that will replace any current subscription
+         * @param timeMs       Time in milliseconds representing the current time
+         */
+        // Visible for testing
+        void updateSubscriptionResult(ClientTelemetrySubscription subscription, long timeMs) {
+            lock.writeLock().lock();
+            try {
+                this.subscription = Objects.requireNonNull(subscription);
+                /*
+                 If the subscription is updated for the client, we want to attempt to spread out the push
+                 requests between 50% and 150% of the push interval value from the broker. This helps us
```

Comment:
this is what the computeStaggeredIntervalMs I assume

Reply:
I find it more appropriate to add the comment here where we are checking if the compute of push interval is required or not. The method comment will mostly talk about what the behaviour of method should be but this comment has additional context. Do let me know if you differ.